frontend/gui_cavity.py

```diff
@@ -0,0 +1,169 @@
+
+
+class GUICavity(BackendCavity):
+    def __init__(self, cavity_num: int, rack_object: "Rack"):
+        super().__init__(cavity_num, rack_object)
+        self._fault_display: Optional[Display] = None
+        self.fault_display_grid_layout = make_header()
+
+        self.vert_layout = QVBoxLayout()
+        self.cavity_widget = CavityWidget()
+        self.cavity_widget.setMinimumSize(10, 10)
+        self.cavity_widget.setAccessibleName("cavity_widget")
+        self.cavity_widget.cavity_text = str(cavity_num)
+        self.cavity_widget.setSizePolicy(QSizePolicy.Preferred, QSizePolicy.Preferred)
+        self.cavity_widget.clicked.connect(self.show_fault_display)
+
+        self.hor_layout = QHBoxLayout()
+
+        self.ssa_bar = PyDMByteIndicator()
+        self.ssa_bar.setAccessibleName("SSA")
+        self.ssa_bar.onColor = QColor(92, 255, 92)
+        self.ssa_bar.offColor = QColor(40, 40, 40)
+        self.ssa_bar.setSizePolicy(QSizePolicy.Maximum, QSizePolicy.Fixed)
+        self.ssa_bar.showLabels = False
+        self.ssa_bar.channel = self.ssa.status_pv
+        self.ssa_bar.setFixedHeight(5)
+
+        self.rf_bar = PyDMByteIndicator()
+        self.rf_bar.setAccessibleName("RFSTATE")
+        self.rf_bar.onColor = QColor(14, 191, 255)
+        self.rf_bar.offColor = QColor(40, 40, 40)
+        self.rf_bar.setSizePolicy(QSizePolicy.Maximum, QSizePolicy.Fixed)
+        self.rf_bar.showLabels = False
+        self.rf_bar.channel = self.rf_state_pv
+        self.rf_bar.setFixedHeight(5)
+
+        self.hor_layout.addWidget(self.ssa_bar)
+        self.hor_layout.addWidget(self.rf_bar)
+
+        self.vert_layout.addWidget(self.cavity_widget)
+        self.vert_layout.addLayout(self.hor_layout)
+
+        severity_pv: str = self.pv_addr("CUDSEVR")
+        status_pv: str = self.pv_addr("CUDSTATUS")
+        description_pv: str = self.pv_addr("CUDDESC")
+
+        self.cavity_widget.channel = status_pv
+        self.cavity_widget.severity_channel = severity_pv
+        self.cavity_widget.description_channel = description_pv
+
+        rule = [
+            {
+                "channels": [
+                    {"channel": self.ssa.status_pv, "trigger": True, "use_enum": True}
+                ],
+                "property": "Opacity",
+                "expression": "ch[0] == 'SSA On'",
+                "initial_value": "0",
+                "name": "show",
+            }
+        ]
+
+        self.ssa_bar.rules = json.dumps(rule)
+
+    def populate_fault_display(self):
+        for idx, fault in enumerate(self.faults.values()):
+            code_label = QLabel(fault.tlc)
+            code_label.setSizePolicy(QSizePolicy.Maximum, QSizePolicy.Maximum)
+            code_label.setAlignment(Qt.AlignCenter)
+
+            short_description_label = QLabel(fault.short_description)
+            short_description_label.setSizePolicy(
+                QSizePolicy.Maximum, QSizePolicy.Maximum
```

Comment:
With the current vertical size policy, the text in short description gets cut off (I've been looking at Hardware Initialization Needed and Hardware Initialization Running for examples) depending on how the fault display is sized. I tested different size policies and it looks like any of the following will work:

Minimum, Preferred, Expanding, MinimumExpanding

We might want to change the line to something like:
`QSizePolicy.Maximum, QSizePolicy.Preferred`